src/mito2/src/region/utils.rs

```diff
@@ -0,0 +1,323 @@
+                path_type,
+            ),
+        ),
+        FileDescriptor::Data(file_id) => (
+            location::sst_file_path(
+                table_dir,
+                RegionFileId::new(source_region_id, file_id),
+                path_type,
+            ),
+            location::sst_file_path(
+                table_dir,
+                RegionFileId::new(target_region_id, file_id),
+                path_type,
+            ),
+        ),
+    }
+}
+
+fn build_delete_file_path(
+    target_region_id: RegionId,
+    file_descriptor: FileDescriptor,
+    table_dir: &str,
+    path_type: PathType,
+) -> String {
+    match file_descriptor {
+        FileDescriptor::Index((file_id, version)) => location::index_file_path(
+            table_dir,
+            RegionIndexId::new(RegionFileId::new(target_region_id, file_id), version),
+            path_type,
+        ),
+        FileDescriptor::Data(file_id) => location::sst_file_path(
+            table_dir,
+            RegionFileId::new(target_region_id, file_id),
+            path_type,
+        ),
+    }
+}
+
+impl RegionFileCopier {
+    pub fn new(access_layer: AccessLayerRef) -> Self {
+        Self { access_layer }
+    }
+
+    /// Copies files from a source region to a target region.
+    ///
+    /// # Arguments
+    ///
+    /// * `source_region_id`: The ID of the source region.
+    /// * `target_region_id`: The ID of the target region.
+    /// * `file_ids`: The IDs of the files to copy.
+    pub async fn copy_files(
+        &self,
+        source_region_id: RegionId,
+        target_region_id: RegionId,
+        file_ids: Vec<FileDescriptor>,
+        parallelism: usize,
+    ) -> Result<()> {
+        ensure!(
+            source_region_id.table_id() == target_region_id.table_id(),
+            InvalidSourceAndTargetRegionSnafu {
+                source_region_id,
+                target_region_id,
+            },
+        );
+        let table_dir = self.access_layer.table_dir();
+        let path_type = self.access_layer.path_type();
+        let object_store = self.access_layer.object_store();
+
+        info!(
+            "Copying {} files from region {} to region {}",
+            file_ids.len(),
+            source_region_id,
+            target_region_id
+        );
+        debug!(
+            "Copying files: {:?} from region {} to region {}",
+            file_ids, source_region_id, target_region_id
+        );
+        let semaphore = Arc::new(Semaphore::new(parallelism));
+        let mut tasks = Vec::with_capacity(file_ids.len());
+        let num_file = file_ids.len();
+        for (index, file_desc) in file_ids.iter().enumerate() {
+            let (source_path, target_path) = build_copy_file_paths(
+                source_region_id,
+                target_region_id,
+                *file_desc,
+                table_dir,
+                path_type,
+            );
+            let moved_semaphore = semaphore.clone();
+            tasks.push(async move {
```

Comment:
We can spawn parallelism tasks and use step_by to reduce the number of async tasks to spawn.

Reply:
[33af457](https://github.com/GreptimeTeam/greptimedb/pull/7389/commits/33af457a2af374dc94d6e9135eee28205bdc13f9)